internal/cri/server/list_pod_sandbox_metrics.go

```diff
@@ -18,12 +18,662 @@ package server
+		log.G(ctx).WithError(err).Error("failed to fetch metrics for task")
+	}
+	if len(resp.Metrics) != 1 {
+		log.G(ctx).Errorf("unexpected metrics response: %+v", resp.Metrics)
+	}
+	cpu, err := c.cpuMetrics(ctx, resp.Metrics[0])
+	if err != nil {
+		log.G(ctx).WithError(err)
+	}
+	sm.metric.Metrics = append(sm.metric.Metrics, generateContainerCPUMetrics(cpu)...)
+
+	memory, err := c.memoryMetrics(ctx, resp.Metrics[0])
+	if err != nil {
+		log.G(ctx).WithError(err)
+	}
+	sm.metric.Metrics = append(sm.metric.Metrics, generateContainerMemoryMetrics(memory)...)
+
+	network, err := c.networkMetrics(ctx, resp.Metrics[0])
+	if err != nil {
+		log.G(ctx).WithError(err)
+	}
+	sm.metric.Metrics = append(sm.metric.Metrics, generateSandboxNetworkMetrics(network)...)
+
+	// get metrics for each container in the sandbox
+	containers := c.containerStore.List()
+	for _, container := range containers {
+		if container.SandboxID == sandboxID {
+			metrics, err := c.listContainerMetrics(ctx, container.ID)
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("failed to list metrics for container %s", container.ID)
+			}
+			sm.metric.ContainerMetrics = append(sm.metric.ContainerMetrics, metrics)
+		}
+	}
+	c.metricsServer.sandboxMetrics[sandboxID] = sm
+	return sm
+}
+
+// getMetrics is supposed to be called from ListPodSandBoxMetrics
+func (m *MetricsServer) getMetrics(sandBoxID string) *runtime.PodSandboxMetrics {
+	var sm *SandboxMetrics
+	// TODO: akhilerm decide if we should query for metrics if this is not available
+	sm, ok := m.sandboxMetrics[sandBoxID]
+	if !ok {
+		// we should not error, but provide the metrics that are available
+	}
+	return sm.metric
+}
+
+func (c *criService) ListPodSandboxMetrics(ctx context.Context, r *runtime.ListPodSandboxMetricsRequest) (*runtime.ListPodSandboxMetricsResponse, error) {
+	sandboxList := c.sandboxStore.List()
+	//metricsList := c.sandboxStore.
+
+	podMetrics := make([]*runtime.PodSandboxMetrics, 0)
+	for _, sandbox := range sandboxList {
+		m := c.metricsServer.getMetrics(sandbox.ID)
+		podMetrics = append(podMetrics, m)
+	}
+
+	return &runtime.ListPodSandboxMetricsResponse{
+		PodMetrics: podMetrics,
+	}, nil
+}
+
+type containerCPUMetrics struct {
+	UsageUsec          uint64
+	UserUsec           uint64
+	SystemUsec         uint64
+	NRPeriods          uint64
+	NRThrottledPeriods uint64
+	ThrottledUsec      uint64
+	//LoadAverage10      uint64
+	//TasksState         uint64
+}
+
+type containerMemoryMetrics struct {
+	Cache        uint64
+	RSS          uint64
+	Swap         uint64
+	KernelUsage  uint64
+	FileMapped   uint64
+	FailCount    uint64
+	MemoryUsage  uint64
+	MaxUsage     uint64
+	WorkingSet   uint64
+	ActiveFile   uint64
+	InactiveFile uint64
+	PgFault      uint64
+	PgMajFault   uint64
+}
+
+type containerNetworkMetrics struct {
+	Name      string
+	RxBytes   uint64
+	RxPackets uint64
+	RxErrors  uint64
+	RxDropped uint64
+	TxBytes   uint64
+	TxPackets uint64
+	TxErrors  uint64
+	TxDropped uint64
+}
```

Comment:
Was thinking if we should add all these into https://github.com/containerd/containerd/blob/main/internal/cri/store/stats/stats.go#L22. cc: @dmcgowan

Reply:
cc: @mikebrow We had earlier discussed to keep this separate, but now thinking whether these fields should be part of https://github.com/containerd/containerd/blob/main/internal/cri/store/stats/stats.go#L22

Reply:
As long as its under internal, it shouldn't matter. These types are just temporary value holders to convert to the api metrics objects? Would they be used anywhere else? We wouldn't want to enlarge structures used internally with values that are only set in one specific case.

Reply:
if you don't need an internal "type" for providing translation between n inputs and m outputs let's not just create one as you'll likely need to know the full details of the source at the time of conversion. IOW don't just create a new cri internal metric type just to have one.  eg. cg1 vs cg2 sources mapping to cri metric going directly allows you to qualify them as cg1 or 2 where nec. if agreement can be made with the crio and kubelet teams .. see networking metrics..

Reply:
https://github.com/containerd/containerd/blob/main/internal/cri/store/stats/stats.go#L22-L27 was needed because the contents ^ were not available from the source metric values to create the summary at the seconds value needed by the rolling summary stats.  A fair to use pattern if necessitate by the client metrics that they specify, or as translation from n sources..

Reply:
IOW.. don't loose information until translation to final.. / don't summarize / normalize unless you have a need to. 

The cri seperates linux/windows .. not sure about other platforms if there is a need for normalization across various versions of linux .. or various windows platforms. Each metric would be an investigation .. what changed are they completely different now.. does the client want both or are they expecting us to convert.

Reply:
once you get into the game of "normalization" forever will it lead you .. and there will always be other consumers who want to know why you threw away details.

Reply:
> once you get into the game of "normalization" forever will it lead you .. and there will always be other consumers who want to know why you threw away details.

But in this case, the network, memory and cpu metrics are from the shim. So we can convert them either into cgroupv1 or v2 and then use those values (not including the windows part here), Rather than converting at the last moment right?

Reply:
we could if we are normalizing the type from the source.. is there a value to that? is that the requirement.. that we normalize first then publish. I've not looked at the requirements for the cri stats.

Reply:
we could if we are normalizing the type from the source (TaskService().Metrics). is there a value to that? is that the requirement.. that we normalize type first then publish. I've not looked at the requirements for the cri stats to determine if we are to normalize first then generate runtime.MetricDescriptor's